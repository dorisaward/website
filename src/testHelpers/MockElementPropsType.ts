import { ReactNode } from 'react'

export type MockElementPropsType<P = unknown> = P & { children?: ReactNode | undefined }
