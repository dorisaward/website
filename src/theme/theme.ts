import { DefaultTheme } from 'styled-components'

export enum theme {
    light = 'LIGHT_THEME',
    dark = 'DARK_THEME',
}

const commonTheme: Pick<DefaultTheme, 'padding'
    | 'switchOnColour'
    | 'switchOffColour'
    | 'shadowColour'
> = {
    padding: '5px',
    switchOnColour: 'green',
    switchOffColour: 'grey',
    shadowColour: 'chocolate',
}

export const lightTheme: DefaultTheme = {
    ...commonTheme,
    backgroundColour: '#CCE6FF',
    secondaryColour: '#FFCCE6',
    tertiaryColour: '#E6FFCC',
    fontColour: 'black'
}

export const darkTheme: DefaultTheme = {
    ...commonTheme,
    backgroundColour: '#214269',
    secondaryColour: '#692142',
    tertiaryColour: '#426921',
    fontColour: 'white'
}
