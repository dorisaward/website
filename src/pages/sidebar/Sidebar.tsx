import styled from 'styled-components'
import { use, useState } from 'react'
import { Outlet } from 'react-router-dom'
import { Switch } from '../../assets/Switch.tsx'
import { theme } from '../../theme/theme.ts'
import { Hamburger } from '../../assets/Hamburger.tsx'
import { LanguagesContext } from '../../languages/LanguagesContext.ts'
import { languages } from '../../languages/languages.ts'
import { Navigation } from '../../assets/Navigation.tsx'
import { MOBILE_WIDTH } from '../../assets/constants.ts'

const SWITCH_WIDTH = '120px'
const HAMBURGER_SIZE = 24

const AppContainer = styled.div`
    position: absolute;
    top: 0;
    left: 0;
    right: 0;
    text-align: center;
    background-color: ${({ theme }) => theme.backgroundColour};
    color: ${({ theme }) => theme.fontColour};
    
    font-family: Inter, system-ui, Avenir, Helvetica, Arial, sans-serif;
    line-height: 1.5;
    font-weight: 400;
    font-synthesis: none;
    text-rendering: optimizeLegibility;
    -webkit-font-smoothing: antialiased;
    -moz-osx-font-smoothing: grayscale;
`

const Container = styled.div<{ isVisible: boolean }>`
    background: ${({ theme }) => theme.secondaryColour };
    display: flex;
    flex-wrap: wrap;
    position: fixed;
    width: 100vw; // mobile-only
    top: 0;
    padding-top: ${HAMBURGER_SIZE + 10}px;
    margin-left: ${({ isVisible }) => (isVisible ? 0 : -100) + '%'};
    transition: margin-left 0.1s;
    @media only screen and (min-width: ${MOBILE_WIDTH}px) {
        width: ${SWITCH_WIDTH};
        margin-left: ${({ isVisible }) => (isVisible ? 0 : '-' + SWITCH_WIDTH)};
    }
`
const HamburgerContainer = styled.div<{ isVisible: boolean }>`
    background: white;
    padding: ${({theme}) => theme.padding};
    width: ${HAMBURGER_SIZE}px;
    height: ${HAMBURGER_SIZE}px;
    position: absolute;
    top: 0;
    margin-left: ${({ isVisible }) => (isVisible ? 0 : 100) + '%'};
    @media only screen and (min-width: ${MOBILE_WIDTH}px) {
        margin-left: ${SWITCH_WIDTH};
    }
`
const ItemContainer = styled.div`
    flex-grow: 1;
    flex-basis: ${SWITCH_WIDTH};
`

export type SidebarProps = {
    theme: {
        selectedTheme: theme
        handleThemePress: () => void
    }
}

export const Sidebar = ({
    theme: { selectedTheme, handleThemePress }
}: SidebarProps) => {
    const [isVisible, setIsVisible] = useState(false)
    const { language, handleLangPress } = use(LanguagesContext)
    const toggleVisibility = () => setIsVisible(prevVisible => !prevVisible)

    const RenderHamburger = () => (
        <HamburgerContainer onClick={toggleVisibility} isVisible={isVisible}>
            <Hamburger
                strokeWidth={4}
                height={HAMBURGER_SIZE}
                width={HAMBURGER_SIZE}
            />
        </HamburgerContainer>
    )

    return (
        <AppContainer>
            <Container isVisible={isVisible}>
                <RenderHamburger />
                <ItemContainer>
                    <Navigation/>
                </ItemContainer>
                <ItemContainer>
                    <Switch
                        checked={selectedTheme === theme.dark}
                        onSwitch={handleThemePress}
                        label={{left: '🌘', right: '☀️'}}
                    />
                    <Switch
                        checked={language === languages.en}
                        onSwitch={handleLangPress}
                        label={{left: 'UK', right: 'TH'}}
                    />
                </ItemContainer>
            </Container>
            <Outlet/>
        </AppContainer>
    )
}