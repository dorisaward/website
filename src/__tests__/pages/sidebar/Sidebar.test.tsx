import { fireEvent, render } from '@testing-library/react'
import { Sidebar, SidebarProps } from '../../../pages/sidebar/Sidebar.tsx'
import { theme } from '../../../theme/theme.ts'
import { MockElementPropsType } from '../../../testHelpers/MockElementPropsType.ts'

jest.mock('react-router-dom', () => ({
    Outlet: ({ children, ...restProps }: MockElementPropsType) => <div {...restProps}>{children}</div>,
}))
const switchPreTestId = 'switchTestId'
jest.mock('../../../assets/Switch.tsx', () => ({
    Switch: ({ children, ...restProps }: MockElementPropsType<{ label: { left: string, right: string } }>) =>
        <div data-testid={switchPreTestId + restProps.label.left} {...restProps}>{children}</div>,
}))
const hamburgerTestId = 'hamburgerTestId'
jest.mock('../../../assets/Hamburger.tsx', () => ({
    Hamburger: ({ children, ...restProps }: MockElementPropsType) =>
        <div data-testid={hamburgerTestId} {...restProps}>{children}</div>,
}))
jest.mock('../../../assets/Navigation.tsx', () => ({
    Navigation: ({ children, ...restProps }: MockElementPropsType) => <div {...restProps}>{children}</div>,
}))

describe('Sidebar', () => {
    it('renders', () => {
        // Given
        const themeProps: SidebarProps['theme'] = {
            selectedTheme: theme.light,
            handleThemePress: jest.fn(),
        }

        // When
        const { baseElement } = render(<Sidebar theme={themeProps}/>)

        // Then
        expect(baseElement).toMatchSnapshot()
    })

    it('hides', () => {
        // Given
        const themeProps: SidebarProps['theme'] = {
            selectedTheme: theme.light,
            handleThemePress: jest.fn(),
        }
        const { getByTestId } = render(<Sidebar theme={themeProps}/>)
        const hamburgerElement = getByTestId(hamburgerTestId)
        const classNameBefore = hamburgerElement.parentElement?.className

        // When
        fireEvent.click(hamburgerElement)

        // Then
        const classNameAfter = getByTestId(hamburgerTestId).parentElement?.className
        expect(classNameBefore).not.toEqual(classNameAfter)
    })
})