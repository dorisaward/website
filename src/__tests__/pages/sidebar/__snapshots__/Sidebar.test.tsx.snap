// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Sidebar renders 1`] = `
<body>
  <div>
    <div
      class="sc-blHHSb bOCNnt"
    >
      <div
        class="sc-gtLWhw clhWGd"
      >
        <div
          class="sc-egkSDF iNTvCT"
        >
          <div
            data-testid="hamburgerTestId"
            height="24"
            stroke-width="4"
            width="24"
          />
        </div>
        <div
          class="sc-fAUdSK eyDKKK"
        >
          <div />
        </div>
        <div
          class="sc-fAUdSK eyDKKK"
        >
          <div
            data-testid="switchTestIdðŸŒ˜"
            label="[object Object]"
          />
          <div
            data-testid="switchTestIdUK"
            label="[object Object]"
          />
        </div>
      </div>
      <div />
    </div>
  </div>
</body>
`;
